doc/cpp_recipe.rst

```diff
@@ -0,0 +1,230 @@
+Recipes
+===============
+
+An Arbor recipe is a description of a model without any information about how the
+model is to be executed. It  provides interface that the model building phase queries
+for information on cells, such as:
+
+  * the number of cells in the model;
+  * the type of a cell;
+  * a description of a cell;
+  * incoming network connections on a cell.
+
+The :cpp:class:`arb::recipe` class documentation is below.
+
+Why Recipes?
+--------------
+
+The Recipe design was motivated by the following aims:
+
+    * Building a simulation from a recipe description must be possible in a
+      distributed system efficiently with minimal communication.
+    * The model description (i.e. recipe) should be seperate from the
+      instantiation of the model to execute on the target system.
+    * To minimise the amount of memory used in model building, to make it
+      possible to build and run simulations in one run.
+
+Recipe descriptions are cell-oriented, in order that the building phase can
+distributed and that the recipe description can be built indepedently of any
+runtime execution environment.
+
+During model building, the recipe is queried first by a load balancer,
+then later when building the low-level cell groups and communication network.
+The cell-centered recipe interface, whereby cell and network properties are
+specified "per-cell", facilitates this.
+
+The steps of building a simulation from a recipe are:
+
+.. topic:: 1. Load balancing
+
+    First, the cell `gid` are partitioned cells over MPI ranks, and each rank
+    the cells to build up a cost model for the cells it has been assigned 
+    Then, the ranks coordinate to redistribute cells over MPI ranks so that
+    each rank has a balanced workload. Finally, each rank groups its local
+    cells into :cpp:type:`cell_group` s that balance the work over threads (and
+    GPU accelerators if available).
+
+.. topic:: 2. Model building
+
+    The model building phase takes the cells assigned to the local rank, and builds the
+    local cell groups and the part of the communication network by querying the recipe
+    for more information about the cells assigned to it.
+
+.. _recipe_best_practice:
+
+Best Practices
+--------------
+
+Here is a set of rules of tumb to keep in mind when making recipes. The first is
+mandatory, and following the others as closely as possible will lead to better
+performance.
+
+.. topic:: Stay thread safe
+
+    The load balancing and model construction are multithreaded, i.e.
```

Comment:
As a general rule, it's usually better to spell out 'i.e.' as 'that is' (and 'e.g' as 'for example' etc.) in longer form text like this.